src/main/java/org/wildfly/security/sasl/entity/EntitySaslServerFactory.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.sasl.entity;
+
+import static org.wildfly.security.sasl.entity.Entity.*;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.Signature;
+import java.util.Map;
+
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslException;
+import javax.security.sasl.SaslServer;
+import javax.security.sasl.SaslServerFactory;
+
+import org.kohsuke.MetaInfServices;
+import org.wildfly.security.sasl.WildFlySasl;
+
+/**
+ * The server factory for the ISO/IEC 9798-3 authentication SASL mechanism.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+@MetaInfServices(value = SaslServerFactory.class)
+public final class EntitySaslServerFactory implements SaslServerFactory {
+
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        Signature signature;
+        boolean mutual = false;
+        final Object serverAuthValue = props.get(Sasl.SERVER_AUTH);
+        final boolean serverAuth = serverAuthValue instanceof String ? "true".equals((String) serverAuthValue) : false;
+        switch (mechanism) {
+            case Entity.ENTITY_MUTUAL_RSA_SHA1_ENC:
+                mutual = true;
+                // Fall through
+            case Entity.ENTITY_UNILATERAL_RSA_SHA1_ENC:
+                if (serverAuth != mutual) return null;
```

Comment:
OK it took me a minute to get what is happening here, so basically we have four conditions:

- SERVER_AUTH is false, unilateral mech selected: OK
- SERVER_AUTH is false, mutual mech selected: refuse mechanism
- SERVER_AUTH is true, unilateral mech selected: refuse mechanism
- SERVER_AUTH is true, mutual mech selected: OK

I wonder if the second case should refuse though.  The definition of ``Sasl.SERVER_AUTH`` is "[...] 'true' if the server must authenticate the to client; 'false' otherwise."  To me this says that a value of 'false' indicates that while mutual auth is not required, it is allowed.  WDYT?